tests/ttnn/unit_tests/operations/test_all_gather_matmul.py

```diff
@@ -9,132 +9,144 @@
+        ag_output_shape, dim, mem_config_ag, num_devices, num_links, ag_input_dtype, layout
     )
     if is_known_failure:
         pytest.skip(f"Skipping unsupported case {message}.")
 
     devices = t3k_device_mesh.get_devices()
 
-    logger.info(f"Input shape: {input_shape}")
+    logger.info(f"All Gather output shape: {ag_output_shape}")
     logger.info(f"dim: {dim}")
 
-    # Create input tensor for the all gather
-    _, _, _, hidden_dim = input_shape
-    input_tensor = torch.rand(input_shape).bfloat16()
+    ##### Create input tensor for the all gather #####
+    _, _, _, hidden_dim = ag_output_shape
+    input_tensor = torch.randn(ag_output_shape).float()
     input_tensors = torch.chunk(input_tensor, num_devices, dim)
     tt_input_tensors = []
     for i, t in enumerate(input_tensors):
-        tt_input_tensors.append(ttl.tensor.Tensor(t, input_dtype).to(layout).to(devices[i], mem_config))
+        tt_input_tensors.append(ttl.tensor.Tensor(t, ag_input_dtype).to(layout).to(devices[i], mem_config_ag))
     input_tensor_mesh = ttnn.aggregate_as_tensor(tt_input_tensors)
 
-    # Create the weight matrix for the matmul
-    weights_tensor = torch.rand([1, 1, hidden_dim, matmul_output_dim * num_devices]).bfloat16()
+    ##### Create the weight matrix for the matmul #####
+    weights_tensor = torch.randn([1, 1, hidden_dim, matmul_output_dim * num_devices]).float()
     weight_tt = ttnn.as_tensor(
         weights_tensor,
-        dtype=input_dtype,
-        layout=ttnn.TILE_LAYOUT,
+        dtype=matmul_weights_dtype,
+        layout=layout,
         device=t3k_device_mesh,
-        memory_config=mem_config,
-        mesh_mapper=ShardTensorToMesh(t3k_device_mesh, dim=3),
+        memory_config=mem_config_weights,
+        mesh_mapper=ShardTensorToMesh(t3k_device_mesh, dim=dim),
     )
 
-    # torch matmul output
-    matmul_output = torch.chunk(torch.matmul(input_tensor, weights_tensor), num_devices, 3)
-
-    # Configs for ttnn.matmul
-    program_config = ttnn.MatmulMultiCoreReuseMultiCast1DProgramConfig(
-        compute_with_storage_grid_size=(1, 1),
-        in0_block_w=16,  # K = 8192 / TILE_WIDTH=32 / Grid_Size is based on compute_with_storage_grid_size
-        out_subblock_h=1,  # Must be divisible by per_core_M
-        out_subblock_w=4,  # Must be divisible by per_core_N, out_subblock_w * out_subblock_h <= 4
-        per_core_M=1,  # M / TILE_HEIGHT = 32 / 32
-        per_core_N=32,  # N / TILE_WIDTH / Grid_Size is based on compute_with_storage_grid_size, N = 4096 for num_device=8
-        fused_activation=None,
-        fuse_batch=True,
-        mcast_in0=True,
-    )
+    ##### Configs for ttnn.matmul #####
+    if matmul_config == "matmul_1d":
+        core_grid = (8, 4)
+        program_config = ttnn.MatmulMultiCoreReuseMultiCast1DProgramConfig(
+            compute_with_storage_grid_size=core_grid,
+            in0_block_w=min(max_in0_block_w, hidden_dim // 32 // core_grid[0]),  # how much inner dim you take each time
```

Comment:
seems there's no test to cover the case `in0_block_w != shard_with`

Reply:
also good to have the test cover `num worker cores != shard grid cores`?

Reply:
Will add a test case where in0_block_w != shard width.

However, for the case where `num worker cores != shard grid cores`, it's currently not supported (I will add a TT_FATAL in the validate). This is because the current implementation is only to support a specific case in LLama, where the output of all gather is sharded across 8 cores.

All gather works by concatenating `n` different tensor slices, each slice contributed by each device (`n=8` devices for t3k).When the shard grid cores = num devices (8), each shard holds exactly a tensor slice. On the matmul side of things, we can benefit from this alignment, because each worker core is responsible for receiving a specific tensor slice (ie, block_id, which is technically shard id, is the same as tensor slice id). So by changing `block_id`, we can essentially select which specific tensor slice to operate on in a custom order.

If a tensor slice was split across multiple cores, then, we would have `num_blocks_per_tensor_slice > num_blocks_per_shard`. This is a problem because the update on [this line](https://github.com/tenstorrent/tt-metal/pull/11760/files/026e1bb962d2b04b92fad28ad84f65dcb4b84132#diff-45cb14f62b525c2ef36d4de6e801edfda0087546281e012256dd4d32e8fa3baeR339) is no longer correct, since block_ids are not aligned with tensor slice ids.

Support for this generality would be added in the future.